### DISCUSSION TITLE
Build DiaNN using Visual Studio 2019

### DISCUSSION THREAD

**Discussion by cjhjhj** [vdemichev/DiaNN#248] [Q&A] [answered]
Hi there,
Is there anyone who could build DiaNN using Visual Studio 2019 in Windows 10? I want to build and go through it step-by-step to figure out the methods/algortihms in DiaNN. However, I couldn't do it. Any suggestion/comment will be appreciated.
Or, any possible way to use something like 'debug' mode which can make me go through the program line-by-line would be great as well.
Thanks.

**Comment by vdemichev** [comment]
You can build versions up to 1.7.12. Please look up the corresponding commits on github, which include Visual Studio project files.

Vadim

**Reply by cjhjhj** [reply]
Hello Vadim, Thanks a lot for your quick response. Let me try 1.7.12 and get back to you soon.

**Reply by cjhjhj** [reply]
I have tried 1.7.12 (and several previous versions like 1.6), but failed to build them. I used Visual Studio 2019, but came across many errors like,
```
E0035	#error directive: The <experimental/filesystem> header providing std::experimental::filesystem is deprecated by Microsoft and will be REMOVED. It is superseded by the C++17 <filesystem> header providing std::filesystem. You can define _SILENCE_EXPERIMENTAL_FILESYSTEM_DEPRECATION_WARNING to acknowledge that you have received this warning.	
Error	C1083	Cannot open include file: 'sys/time.h': No such file or directory	diann	C:\Software\DiaNN-1.7.12\mstoolkit\src\expat-2.2.0\xmlparse.c	13	
Error (active)	E1696	cannot open source file "C:/Software/DiaNN-1.7.12/diann/Debug/XRawfile2_x64.tlh"	diann	C:\Software\DiaNN-1.7.12\mstoolkit\include\MSReader.h	39	
Error (active)	E1696	cannot open source file "C:/Software/DiaNN-1.7.12/diann/Debug/XRawfile2_x64.tlh"	diann	C:\Software\DiaNN-1.7.12\mstoolkit\include\RAWReader.h	34	
Error	C1083	Cannot open source file: '..\sqlite\sqlite3.c': No such file or directory	diann	C:\Software\DiaNN-1.7.12\diann\c1	1	
Error (active)	E0020	identifier "_set_FMA3_enable" is undefined	diann	C:\Software\DiaNN-1.7.12\src\diann.cpp	10898	
Error (active)	E0725	name must be a namespace name	diann	C:\Software\DiaNN-1.7.12\mstoolkit\include\MSReader.h	41	
Error (active)	E0725	name must be a namespace name	diann	C:\Software\DiaNN-1.7.12\mstoolkit\include\RAWReader.h	36
```	
I could build the very first version using Visual Studio 2019, but failed for any other versions. Could you please give me some advice?
Thank you.

**Reply by vdemichev** [reply]
Please see the generic docs on how to troubleshoot these specific errors, maybe compiler settings need to be changed. It's definitely possible to build 1.7.12.

**Reply by vdemichev** [reply]
Also, probably need to install Thermo MSFileReader (the specific version recommended for DIA-NN)

**Reply by cjhjhj** [reply]
Thank you so much for your quick reply. Let me try to troubleshoot.

**Reply by cjhjhj** [reply]
Sorry for keep bothering you, but I still have a problem with building 1.7.12 with the following error,
```
Error C1083 Cannot open include file: 'sys/time.h': No such file or directory | diann | C:\Software\DiaNN-1.7.12\mstoolkit\src\expat-2.2.0\xmlparse.c
```
It would be great if you could give me advice about the error.

</body>
</html>

**Reply by vdemichev** [reply]
Apologies, but I don't really remember much about all this. Can say that it should be OK to build straight away, strange that these errors pop up.

**Reply by cjhjhj** [reply]
Thank you for the comments. Unfortunately, due to the error, the build cannot be done in VS2019. Let me try some other ways, anyway.

**Comment by cjhjhj** [comment, marked as answer]
Hi Vadim,
Finally, I could build the source and get into debug mode. I want to share my troubleshoot just for information.
- Vdproj unsupported
    - It can be resolved by installing an extension (https://stackoverflow.com/questions/26028746/visual-studio-vdproj-is-incompatible)
- Try to build as is (no change any settings)
    - Additional include directory needs to be added (/mstoolkit/include)
- /sys/time.h related error
    - Project properties -> Configuration properties -> C/C++ -> Command line: if not, add an option /D "WIN32"  
- /sqlite/sqlite3.c related error
    - Make a directory /sqlite (under the project directory) and copy splite3.c (from /mstoolkt/src/sqlitexxx) to the directory
- _SILENCE_EXPERIMENTAL_FILESYSTEM_DEPRECATION_WARNING related error
    - Project properties -> Configuration properties -> C/C++ -> Preprocesser: add _SILENCE_EXPERIMENTAL_FILESYSTEM_DEPRECATION_WARNING
- _set_FMA3_enable(0) related error
    - Comment the line
- Link error: mismatch related with _ITERATOR_DEBUG_LEVEL
    - Changed the diann.cpp file as follows,
      #define _HAS_ITERATOR_DEBUGGING 1 
      #define _ITERATOR_DEBUG_LEVEL 2

Thanks.

**Reply by vdemichev** [reply]
Thank you!
